From dd455606e9a48852d2d4958100ec351d49cbb4e7 Mon Sep 17 00:00:00 2001
From: BuddyZhang1 <buddy.zhang@aliyun.com>
Date: Tue, 5 Mar 2019 20:21:49 +0800
Subject: [PATCH 1/1] add debug function for memblock_reserved

---
 linux-5.0/arch/arm/kernel/setup.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/linux-5.0/arch/arm/kernel/setup.c b/linux-5.0/arch/arm/kernel/setup.c
index 375b13f..d36d824 100644
--- a/linux-5.0/arch/arm/kernel/setup.c
+++ b/linux-5.0/arch/arm/kernel/setup.c
@@ -1074,6 +1074,10 @@ void __init setup_arch(char **cmdline_p)
 {
 	const struct machine_desc *mdesc;
 
+#ifdef CONFIG_BISCUITOS_MEMBLOCK_RESERVE
+	extern int debug_memblock_reserve(void);
+#endif
+
 	setup_processor();
 	mdesc = setup_machine_fdt(__atags_pointer);
 	if (!mdesc)
@@ -1104,6 +1108,10 @@ void __init setup_arch(char **cmdline_p)
 	strlcpy(cmd_line, boot_command_line, COMMAND_LINE_SIZE);
 	*cmdline_p = cmd_line;
 
+#ifdef CONFIG_BISCUITOS_MEMBLOCK_RESERVE
+	debug_memblock_reserve();
+#endif
+
 	early_fixmap_init();
 	early_ioremap_init();
 
-- 
2.7.4

